extension/opampextension/registry.go

```diff
@@ -0,0 +1,205 @@
+	callbackElem := capabilityList.PushBack(msgChan)
+
+	unregisterFunc := cr.removeCapabilityFunc(capability, callbackElem)
+	sender := newCustomMessageSender(cr, cr.client, capability, msgChan, unregisterFunc)
+
+	return sender, nil
+}
+
+// ProcessMessage processes a custom message, asynchronously broadcasting it to all registered capability handlers for
+// the messages capability.
+func (cr customCapabilityRegistry) ProcessMessage(cm *protobufs.CustomMessage) {
+	cr.mux.Lock()
+	defer cr.mux.Unlock()
+
+	msgChannels, ok := cr.capabilityToMsgChannels[cm.Capability]
+	if !ok {
+		return
+	}
+
+	for node := msgChannels.Front(); node != nil; node = node.Next() {
+		msgChan, ok := node.Value.(chan *protobufs.CustomMessage)
+		if !ok {
+			continue
+		}
+
+		// If the channel is full, we will skip sending the message to the receiver.
+		// We do this because we don't want a misbehaving component to be able to
+		// block the opamp extension, or block other components from receiving messages.
+		select {
+		case msgChan <- cm:
+		default:
+		}
+	}
+}
+
+// removeCapabilityFunc returns a func that removes the custom capability with the given msg channel list element and sender,
+// then recalculates and sets the list of custom capabilities on the OpAMP client.
+func (cr *customCapabilityRegistry) removeCapabilityFunc(capability string, callbackElement *list.Element) func() {
+	return func() {
+		cr.mux.Lock()
+		defer cr.mux.Unlock()
+
+		msgChanList := cr.capabilityToMsgChannels[capability]
+		msgChanList.Remove(callbackElement)
+
+		if msgChanList.Front() == nil {
+			// Since there are no more callbacks for this capability,
+			// this capability is no longer supported
+			delete(cr.capabilityToMsgChannels, capability)
+		}
+
+		capabilities := cr.capabilities()
+		err := cr.client.SetCustomCapabilities(&protobufs.CustomCapabilities{
+			Capabilities: capabilities,
+		})
+		if err != nil {
+			// It's OK if we couldn't actually remove the capability, it just means we won't
+			// notify the server properly, and the server may send us messages that we have no associated callbacks for.
+			cr.logger.Error("Failed to set new capabilities", zap.Error(err))
+		}
+	}
+}
+
+// capabilities gives the current set of custom capabilities with at least one
+// callback registered.
+func (cr *customCapabilityRegistry) capabilities() []string {
+	return maps.Keys(cr.capabilityToMsgChannels)
+}
+
+type customMessageHandler struct {
+	// unregisteredMux protects unregistered, and makes sure that a message cannot be sent
+	// on an unregistered capability.
+	unregisteredMux *sync.Mutex
+
+	capability               string
+	opampClient              customCapabilityClient
+	registry                 *customCapabilityRegistry
+	sendChan                 <-chan *protobufs.CustomMessage
+	unregisterCapabilityFunc func()
+
+	unregistered bool
+}
+
+var _ CustomCapabilityHandler = (*customMessageHandler)(nil)
+
+func newCustomMessageSender(
```

Comment:
This should be `newCustomMessageHandler`.

Reply:
Whoops, fixed.